### DISCUSSION TITLE
Something runs out

### DISCUSSION THREAD

**Discussion by mtibi75** [arendst/Tasmota#21374] [Support request]
Hi
I developed a large program, it's about 1000 lines in Berry.
It's working fine, everything is OK.
BUT, something runs out, i cant add a new var or a new def, the program crashes.
I only add: Var x and i get a completely meaningless, irrelevant error message, it has nothing to do with the variable x.

Do you have any idea what am I spoiling?
[info.txt](https://github.com/arendst/Tasmota/files/15221715/info.txt)
Thanks, mtibi75

**Comment by sfromis** [comment]
Unless having more details than "something runs out", it is too much guesswork. Even an "irrelevant" message can be a clue about what happens.

Still, if you have a low-memory situation, than can very much be a risk factor for unpredictable crashes.

**Comment by mtibi75** [comment]
Now, the sendmail function is skipped, if I enable it, it doesn't work anymore. So I'm at the limit.
So it is stable and works for days.
"Version":"13.4.1.2(TasmoCompiler-esp32s3)","BuildDateTime":"2024-04-28T07:43:44","Core":"3_0_0","SDK":"5.1.3.240325","CpuFrequency":240,"Hardware":"ESP32-S3 v0.2","CR":"570/699"

tasmota.memory(): {'stack_low': 17.8086, 'program_free': 752, 'program': 2127, 'psram': 2048, 'stack_size': 24, 'flash_real': 8192, 'iram_free': 0, 'flash': 8192, 'frag': 6, 'heap_free': 212, 'psram_free': 1699}
tasmota.get_free_heap(): 217212
"Berry":{"HeapUsed":126,"Objects":1755}
I don't know if the StackLowMark, iramfree is too small? Or Berry's object have got limit? because the other values seem to be enough.

**Reply by sfromis** [reply]
Using SendMail is pretty heavy function, IIRC needing quite a bit of available RAM. Memory numbers only become "interesting" if taken at the time it would croak.....
